---
title: "Britt RNA-Seq Analysis: ENSEMBL Gene Level"
author: "Lauren Vanderlinden"
date: "February 20, 2018"
output: html_document
---

### Overview ###

**Experimental Design**

This is a classic 2x2 design.  The 2 variables of interest are strain and tissue.  The strains are white eyed (W) and sevenless (S) and the tissues we are interested in are optic lope (O) and retina (R).  We really want to get at strain effects and categorize the different types of strain effects (e.g. strains only different within retina and not expressed at all in optic lobe).  There are 4 sample types and four replicates, SO#, SR#, WO#, WR#.

**Pre-processing**

The first part of this document goes over the pre-processing and normalization.  It records how the RNA-sequencing data is processed starting from RSEM results to data used for statistical analyses.    

**Analysis**

The second part of this document goes over the statistical modeling.  We will first look at an overall strain effect by performing a likelihood ratio test (LRT) between the full model (expression = strain + tissue + strain*tissue) vs the null model (expression = tissue).  Then we will separate out the significant strain effects into various categories.  

### Pre-processing Overview ###

**1.** Load gene-level RSEM expected counts

**2.** Remove genes not expressed above background

**3.** Examine data for outliers

### Data ###

The raw data was downloaded on **November 6, 2017** and saved on sysgen under: ```/data/home/sabal/Britt/RNA-Seq.2017-11-06/rawReads```

The trimmed files 
(cudadapt v1.9.1, code:```/data/home/vanderll/Britt/RNAseq.20171106/code/trimmReads.v3.sh```)
are located under:
```/data/home/vanderll/Britt/RNAseq.20171106/trimmedReads```

The aligned files to the dm6 genome
(hisat2 v2.0.3, code:```python /data/usr/local/alignBatch.genome.hisat.py --input-suffix .fastq.gz --index-dir /data/home/vanderll/annotation/dm6.hisat2.reference -P /data/home/vanderll/Britt/RNAseq.20171106/trimmedReads```)
are located under:
```/data/home/vanderll/Britt/RNAseq.20171106/alignedReads```

The RSEM quantitated files 
(RSEM v1.2.31, code:```python runRSEM_batch.py --rsem-time --rsem-seedLen 20 --rsem-seed 2020 --rsem-bowtie2 --rsem-noBam --rsem-fwProb 0.0 --paired -d _R -o /data/home/vanderll/Britt/RNAseq.20171106/RSEMquant.ENSEMBL/ -i trimmed.fastq.gz /data/home/sabal/Britt/RNA-Seq.2017-11-06/trimmedReads/ /data/home/vanderll/Britt/RNAseq.20171106/index/ dm6.ensembl 8```)
are located under: ```/data/home/vanderll/Britt/RNAseq.20171106/RSEMquant.ENSEMBL/```  

Overview of data: they are pair-ended reads with a read length of 150.  Trimmed using cutadapt and after examining fastQC files, the first 15 bases need to be trimmed off the reads. This gives us a read length of 135.  The trimmed files were then aligned to the dm6 genome (using hisat2) as well as quantified to the Ensembl dm6 transcriptome (RSEM).

Please see the ```/data/home/vanderll/Bitt/RNAseq.20171106/code/RNAseq.preprocessing``` file for specific code and details on all the pre-processing steps done outside of R.  

```{r, warning=FALSE, message=FALSE, error=FALSE, cache=FALSE, include=FALSE}
#set up workspace
rm(list=ls())
options(stringsAsFactors=FALSE)

library(limma)
library(DESeq2) 
library(DESeq) 
library(RUVnormalize)
require(ggplot2)
library(RColorBrewer)
library(RUVSeq)
library(WGCNA)
library(knitr)
library(EDASeq)
library(RColorBrewer)
library(edgeR)
#library(enrichR)


```

```{r, warning=FALSE, message=FALSE, error=FALSE, echo=FALSE}
#get the file names and examine the # strains, any technical replicates and so on. 
wd = "Y:/Britt/RNAseq.20171106/RSEMquant.ENSEMBL/"

getGeneResults = function(a){
  b = a[grep(".gene",a)]
  return(b)
}

files =  paste(wd, getGeneResults(list.files(wd)), sep="")


getSample = function(a){
  b = sapply(strsplit(sapply(strsplit(a, ".genes", fixed=TRUE), "[[", 1), split="/", fixed=TRUE), "[[", 5)
  return(b)
}

samples = getSample(files)
files.v2 = as.data.frame(cbind(files, samples))
colnames(files.v2) = c("file", "sample")

group = c()
group[grep("_SO", files.v2$sample)] = "SO"
group[grep("_SR",files.v2$sample)] = "SR"
group[grep("_WO",files.v2$sample)] = "WO"
group[grep("_WR",files.v2$sample)] = "WR"
files.v2$group = group

tissue = c()
tissue[grep("_SO", files.v2$sample)] = "opticLobe"
tissue[grep("_SR",files.v2$sample)] = "retina"
tissue[grep("_WO",files.v2$sample)] = "opticLobe"
tissue[grep("_WR",files.v2$sample)] = "retina"
files.v2$tissue = tissue

strain = c()
strain[grep("_S", files.v2$sample)] = "sevenless"
strain[grep("_W",files.v2$sample)] = "whiteEyed"
files.v2$strain = strain
```

```{r, warning=FALSE, message=FALSE, error=FALSE, eval=FALSE, echo=FALSE}
#load in the data
for(i in 1:nrow(files.v2)){
  x = read.table(file=files.v2[i,"file"],sep="\t",header=TRUE)
  x = x[,c("gene_id","expected_count")]
  colnames(x)[2] = files.v2[i, "sample"]
  if(files.v2[i,"file"]!=files.v2[1, "file"]) rsem = merge(x,rsem,by=c("gene_id"),all=TRUE)
  if(files.v2[i, "file"]==files.v2[1, "file"]) rsem = x
}

x = read.table(file=files.v2[i,"file"],sep="\t",header=TRUE)

estCnts = rsem[,-1]
rownames(estCnts) = rsem$gene_id

counts = round(estCnts,0)
save(counts, file="Y:/Britt/RNAseq.20171106/data/RSEM.ensembl.estCounts.noFiltering.Rdata")
```
```{r, warning=FALSE, message=FALSE, error=FALSE, echo=FALSE}
load(file="Y:/Britt/RNAseq.20171106/data/RSEM.ensembl.estCounts.noFiltering.Rdata")
```


```{r, warning=FALSE, message=FALSE, error=FALSE, eval=FALSE, echo=FALSE}
wd = "Y:/Britt/RNAseq.20171106/RSEMquant.ENSEMBL/"

getTranscriptResults = function(a){
  b = a[grep(".isoform",a)]
  return(b)
}

files =  paste(wd, getTranscriptResults(list.files(wd)), sep="")

#load in the data
for(i in 1:length(files)){
  x = read.table(file=files[i],sep="\t",header=TRUE)
  x = x[,c("transcript_id","expected_count")]
  #colnames(x)[2] = files.v2[i, "sample"]
  if(files[i]!=files[1]) rsem = merge(x,rsem,by=c("transcript_id"),all=TRUE)
  if(files[i]==files[1]) rsem = x
}


estCnts = rsem[,-1]
rownames(estCnts) = rsem$transcript_id

transcript.counts = round(estCnts,0)
save(transcript.counts, file="Y:/Britt/RNAseq.20171106/data/RSEM.ensembl.transcript.estCounts.noFiltering.Rdata")
```
```{r, warning=FALSE, message=FALSE, error=FALSE, echo=FALSE}
load(file="Y:/Britt/RNAseq.20171106/data/RSEM.ensembl.transcript.estCounts.noFiltering.Rdata")
```

### Sample Counts ###

Looking at the sum of reads in each sample at each point in the data processing.  Please note for the hisat2 alignment, there are many types of alignment (e.g. paired reads aligning coordinately or discordinately, or only 1 pair aligning, how many times...), so I just reported overall alignment as a percentage.    

```{r, warning=FALSE, message=FALSE, error=FALSE, echo=FALSE}

### raw reads
raw = read.table(file="Y:/Britt/RNAseq.20171106/data/rawReadCounts.txt")
raw$sample = sapply(strsplit(sapply(strsplit(raw$V1, split="_R", fixed=TRUE), "[[", 1), split="rawReads/", fixed=TRUE), "[[", 2)

raw2 = raw[,c("V3", "sample")]
colnames(raw2)[1] = "raw total Counts"
raw2 = raw2[!duplicated(raw2),]

### trimmed reads
trimmed = read.table(file="Y:/Britt/RNAseq.20171106/data/trimmedReadCounts_try2.txt")
trimmed$sample = sapply(strsplit(sapply(strsplit(trimmed$V1, split="_R", fixed=TRUE), "[[", 1), split="trimmedReads/", fixed=TRUE), "[[", 2)

trimmed2 = trimmed[,c("V3", "sample")]
colnames(trimmed2)[1] = "trimmed total Counts"
trimmed2 = trimmed2[!duplicated(trimmed2),]

###aligned reads
aligned = read.csv(file="Y:/Britt/RNAseq.20171106/data/alignedInfo.csv")
aligned$sample = paste(aligned$sample, "_S", c(1:16), sep="")
colnames(aligned)[2] = "Overall Alignment"
###rsem reads
rsem.sums = as.matrix(colSums(counts))
colnames(rsem.sums) = "RSEM total Counts"

m1 = merge(raw2, trimmed2, by="sample")
m2 = merge(m1, rsem.sums, by.x="sample", by.y=0)
m2$trimmPercent = (m2$`trimmed total Counts`/m2$raw)*100
m2$rsemPercent.fromTrimm = (m2$`RSEM total Counts`/m2$`trimmed total Counts`)*100

colnames(m2) = c("sample", "raw", "trim", "RSEM", "trim percent", "RSEM percent from trimm")
m2[,2] = prettyNum(m2[,2], big.mark=",")
m2[,3] = prettyNum(m2[,3], big.mark=",")
m2[,4] = prettyNum(m2[,4], big.mark=",")
m2[,5] = paste(round(m2[,5], 2), "%", sep="")
m2[,6] = paste(round(m2[,6], 2), "%", sep="")

m3 = merge(m2, aligned, by="sample")
m3 = m3[,c(1:5, 7,6)]

orderWant = names(rsem.sums[(order(rsem.sums)),])
orderGet = c()
for(i in orderWant){
  orderGet = c(orderGet, which(m3$sample==i))
}
m3 = m3[orderGet,]
```

The following table samples are reported in ascending order of RSEM counts.

`r kable(m3, align="c", row.names=FALSE)`

**Look at Distribution of Genes**

```{r, message=FALSE, warning=FALSE, eval=FALSE,echo=FALSE}
test = read.delim(file="Y:/annotation/Drosophila_melanogaster.BDGP6.88.LVupdatedChr.gtf",header=FALSE)

#get gene annotation;
test.gene = test[which(test$V3=="gene"),]
gene.id = gsub("gene_id ", "", sapply(strsplit(test.gene$V9, split=";", fixed=TRUE), "[[", 1))
gene.symbol = gsub(" gene_name ", "", sapply(strsplit(test.gene$V9, split=";", fixed=TRUE), "[[", 2))
type = gsub(" gene_biotype ", "", sapply(strsplit(test.gene$V9, split=";", fixed=TRUE), "[[", 4))
chr = test.gene$V1
start = test.gene$V4
stop = test.gene$V5
strand = test.gene$V7

gene.anno = data.frame(gene.id, gene.symbol, chr, start, stop, strand, type)

#get transcript annotation;
test.transcript = test[which(test$V3=="transcript"),]
transcript.id = gsub(" transcript_id ", "", sapply(strsplit(test.transcript$V9, split=";", fixed=TRUE), "[[", 2))
gene.id = gsub("gene_id ", "", sapply(strsplit(test.transcript$V9, split=";", fixed=TRUE), "[[", 1))
gene.symbol = gsub(" gene_name ", "", sapply(strsplit(test.transcript$V9, split=";", fixed=TRUE), "[[", 3))
transcript.name = gsub(" transcript_name ", "", sapply(strsplit(test.transcript$V9, split=";", fixed=TRUE), "[[", 6))
type = gsub(" transcript_biotype ", "", sapply(strsplit(test.transcript$V9, split=";", fixed=TRUE), "[[", 8))
chr = test.transcript$V1
start = test.transcript$V4
stop = test.transcript$V5
strand = test.transcript$V7

transcript.anno = data.frame(transcript.id, gene.id, transcript.name, gene.symbol, chr, start, stop, strand, type)

save(gene.anno, transcript.anno, file="Y:/annotation/anno.BDGP6.88.bothGeneAndTranscript.Rdata")
```
```{r, message=FALSE, warning=FALSE,echo=FALSE}
load(file="Y:/annotation/anno.BDGP6.88.bothGeneAndTranscript.Rdata")

geneType2 = prettyNum(as.matrix(table(gene.anno$type)), big.mark=",")
proportion = paste(round(as.matrix(table(gene.anno$type))/nrow(gene.anno)*100, 2), "%", sep="")

geneType2 = cbind(geneType2, proportion)
colnames(geneType2)[1] = "Number of Genes"
rownames(geneType2) = names(table(gene.anno$type))

gene.anno.want = gene.anno[which(gene.anno$gene.id %in% as.matrix(rownames(counts))),]
#table(gene.anno.want$gene.id==rownames(counts))
#get right order;




proportion.Reads = c()
for(i in rownames(geneType2)){
  tmp.anno = gene.anno.want[which(gene.anno.want$type==i),]
  tmp = counts[which(rownames(counts) %in% tmp.anno$gene.id),]
  sum.reads = sum(tmp)
  proportion.Reads = c(proportion.Reads, paste(round(sum(tmp)/sum(counts)*100, 5), "%", sep=""))
}

geneType2 = cbind(geneType2, proportion.Reads)
colnames(geneType2)[2:3] = c("Proportion of Gene Type", "Proportion of RSEM Counts")
###
transcriptType2 = prettyNum(as.matrix(table(transcript.anno$type)), big.mark=",")
proportion = paste(round(as.matrix(table(transcript.anno$type))/nrow(transcript.anno)*100, 2), "%", sep="")

transcriptType2 = cbind(transcriptType2, proportion)
colnames(transcriptType2)[1] = "Number of Transcripts"
rownames(transcriptType2) = names(table(transcript.anno$type))

transcript.anno.want = transcript.anno[which(transcript.anno$gene.id %in% rownames(counts)),]

proportion.Reads = c()
for(i in rownames(transcriptType2)){
  tmp.anno = transcript.anno.want[which(transcript.anno.want$type==i),]
  tmp = counts[which(rownames(counts) %in% tmp.anno$gene.id),]
  sum.reads = sum(tmp)
  proportion.Reads = c(proportion.Reads, paste(round(sum(tmp)/sum(counts)*100, 5), "%", sep=""))
}

transcriptType2 = cbind(transcriptType2, proportion.Reads)
colnames(transcriptType2)[2:3] = c("Proportion of Transcript Type", "Proportion of RSEM Counts")



```

There are a total of **`r prettyNum(nrow(gene.anno), big.mark=",")`** genes in the fly Ensembl transcriptome.  

`r kable(geneType2, align="c")`

The RNA-Seq looks to be polyA selected as the majority of reads are coming from the protein coding.  
